tests/test_tools.py

```diff
@@ -16,109 +16,201 @@ def _setup_dirs(monkeypatch, user_filesystem):
     return home_dir
 
 
-def _run_tests(inputs, expected):
-    args = {"username": inputs[0], "email": inputs[1]}
-    expected_username, expected_email = expected
-    config = get_user_info(args)
+def _run_tests(cli_inputs, expected):
+    user_info = {"username": cli_inputs["cli_username"], "email": cli_inputs["cli_email"]}
+    config = get_user_info(user_info=user_info, skip_config_creation=cli_inputs["skip_config_creation"])
+    expected_username = expected["expected_username"]
+    expected_email = expected["expected_email"]
     assert config.get("username") == expected_username
     assert config.get("email") == expected_email
 
 
 params_user_info_with_home_conf_file = [
-    (["", ""], ["home_username", "home@email.com"]),
-    (["cli_username", ""], ["cli_username", "home@email.com"]),
-    (["", "cli@email.com"], ["home_username", "cli@email.com"]),
-    ([None, None], ["home_username", "home@email.com"]),
-    (["cli_username", None], ["cli_username", "home@email.com"]),
-    ([None, "cli@email.com"], ["home_username", "cli@email.com"]),
-    (["cli_username", "cli@email.com"], ["cli_username", "cli@email.com"]),
+    (
+        {"cli_username": None, "cli_email": None, "skip_config_creation": False},
+        {"expected_username": "home_username", "expected_email": "home@email.com"},
+    ),
+    (
+        {"cli_username": "cli_username", "cli_email": None, "skip_config_creation": False},
+        {"expected_username": "cli_username", "expected_email": "home@email.com"},
+    ),
+    (
+        {"cli_username": None, "cli_email": "cli@email.com", "skip_config_creation": False},
```

Comment:
Hey Bob, thanks for the feedback!! I think these repeated key values correspond to different test cases. Basically this function tries to get username/email from four places: (a) prompt input, (b) function arguments (which we called cli input because it was initially used by labpdfproc, where the function arguments are from cli), (c) local config file, (d) global config file, with priority order.

Currently we have four test functions that correspond to:
(1) we only have global config file, function prioritizes (b) to (d) (that is, replace (d) if (b) is non-empty).
(2) we have local config file with/without global config file, function prioritizes (b) to (c).
(3) we dont have config files, and user does not choose to skip config creation, then function prioritizes (a) to (b), and create a global config file if both username/email arguments are present.
(4) we dont have config files, and user chooses to skip config creation (this means we don't have (a)), then the function reads (b), and replace username/email with an empty string if no value given.

So I think all arguments and test values are necessary here. I will add a bit more comment to the tests so that it's clearer. We might be able to find a better way to organize the tests?

Reply:
Thanks @yucongalicechen for the comments.  Yeah I had a quick look at the src code. 

Username/email from 4 places:

1. prompt
2. function args
3. local config
4. global config

and the test functions are basically testing the input/out and reading/creating config files as needed.. 

Noted.